File: games/rdr2/honor_actions.py
"""Exact, editor-facing model of Story Mode's shared honor controls (#62).

Rockstar's Story scripts use a global event-block bitmask and a fixed 19-entry
magnitude function.  They do not provide one editable value per named action:
call sites choose one of these shared tiers, often with the same event hash.
"""
from __future__ import annotations

import csv
import os
import tempfile
from pathlib import Path


EVENTS = (
    ("HONOR_EVENT_LOOT_INNOCENT", 1, "Loot an innocent"),
    ("HONOR_EVENT_AMBIENT_KILL", 2, "Kill a person"),
    ("HONOR_EVENT_AMBIENT_KO", 4, "Knock out a person"),
    ("HONOR_EVENT_SCARE", 32, "Scare a person"),
    ("HONOR_EVENT_KILL_VERMIN", 64, "Kill vermin"),
    ("HONOR_EVENT_KILL_FARM_ANIMAL", 128, "Kill a farm animal or dog"),
    ("HONOR_EVENT_KILL_HORSE", 256, "Kill a horse"),
    ("HONOR_EVENT_STEAL_HORSE", 512, "Steal a horse"),
    ("HONOR_EVENT_STEAL_DONKEY", 1024, "Steal a donkey"),
    ("HONOR_EVENT_STEAL_MULE", 2048, "Steal a mule"),
    ("HONOR_EVENT_TRAMPLED_INNOCENT", 4096, "Trample an innocent"),
    ("HONOR_EVENT_STEAL_WAGON", 8192, "Steal a wagon"),
    ("HONOR_EVENT_ABANDON_ANIMALS", 16384, "Abandon hunted animals"),
    ("HONOR_EVENT_ANIMAL_BLEEDOUT", 32768, "Let an animal bleed out"),
    ("HONOR_EVENT_ANTAGONIZE", 65536, "Antagonize"),
    ("HONOR_EVENT_THEFT", 131072, "Theft"),
    ("HONOR_EVENT_INTERVENED", 262144, "Intervene or help"),
    ("HONOR_EVENT_WANTED_IN_CAMP", 524288, "Bring law or combat into camp"),
    ("HONOR_EVENT_DONATED_GAME", 1048576, "Donate game"),
    ("HONOR_EVENT_ITEM_REQUEST", 2097152, "Fulfil an item request"),
    ("HONOR_EVENT_LONG_ABSENCE", 4194304, "Return after a long absence"),
)

TIERS = (-640, -480, -320, -160, -40, -20, -10, -5, -2, -1,
         0, 1, 2, 5, 10, 20, 40, 160, 640)


def _defaults() -> dict:
    return {
        "events": [{"id": key, "bit": bit, "label": label, "enabled": True}
                   for key, bit, label in EVENTS],
        "tiers": [{"id": f"tier_{value:+d}", "vanilla": value,
                   "amount": value, "enabled": True} for value in TIERS],
    }


def read_honor_actions(path: Path) -> dict:
    data = _defaults()
    if path.exists():
        rows = list(csv.DictReader(path.open(encoding="utf-8-sig", newline="")))
        event_by_id = {row["id"]: row for row in data["events"]}
        tier_by_id = {row["id"]: row for row in data["tiers"]}
        for row in rows:
            target = event_by_id.get(row.get("id")) or tier_by_id.get(row.get("id"))
            if not target:
                raise ValueError(f"unknown honor control: {row.get('id')!r}")
            target["enabled"] = row.get("enabled", "1").strip().lower() not in ("0", "false", "no")
            if "amount" in target:
                target["amount"] = int(row["amount"])
    data.update({
        "available": True,
        "file": str(path),
        "scopeNote": "Edit replacement amounts in the honor-amount table. Event toggles map to Rockstar's exact Global_36616 bits; amounts are shared tiers selected by script call sites, not independent values for each event/action.",
        "bountyAudit": "Hostile human bounty hunters are recognized by REL_BOUNTY_HUNTER. PoliceDog uses A_C_DogHound_01, but short_update classifies all dog animal types as farm animals and applies HONOR_EVENT_KILL_FARM_ANIMAL unless that ped is explicitly blocked.",
    })
    return data


def save_honor_actions(path: Path, edits: list[dict]) -> int:
    data = read_honor_actions(path)
    controls = {row["id"]: row for row in data["events"] + data["tiers"]}
    seen = set()
    for edit in edits:
        key = str(edit.get("id", ""))
        if key in seen or key not in controls:
            raise ValueError(f"duplicate or unknown honor control: {key!r}")
        seen.add(key); row = controls[key]
        if "enabled" in edit:
            value = edit["enabled"]
            if not isinstance(value, bool):
                raise ValueError(f"enabled must be boolean for {key}")
            row["enabled"] = value
        if "amount" in edit:
            if "amount" not in row:
                raise ValueError(f"honor event {key} has no independent amount")
            row["amount"] = int(edit["amount"])
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, temporary = tempfile.mkstemp(prefix=path.name + ".", suffix=".tmp", dir=path.parent)
    with os.fdopen(fd, "w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=("kind", "id", "enabled", "amount"))
        writer.writeheader()
        for row in data["events"]:
            writer.writerow({"kind": "event", "id": row["id"], "enabled": int(row["enabled"]), "amount": ""})
        for row in data["tiers"]:
            writer.writerow({"kind": "tier", "id": row["id"], "enabled": int(row["enabled"]), "amount": row["amount"]})
    os.replace(temporary, path)
    # Strict round trip catches malformed or silently dropped settings.
    check = read_honor_actions(path)
    if len(check["events"]) != len(EVENTS) or len(check["tiers"]) != len(TIERS):
        raise ValueError("honor controls failed round-trip validation")
    return len(edits)

File: games/rdr2/native_runtime/GameplayTweaks/modules/honor_actions.cpp
// GitHub #62: exact Story honor-event bit controls and shared tier remapping.
// Included by the integration dispatcher after common wrappers are defined.

struct HonorActionTier { int vanilla; int amount; bool enabled; };

static const int kHonorEventBits[] = {
	1, 2, 4, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384,
	32768, 65536, 131072, 262144, 524288, 1048576, 2097152, 4194304
};
static const char* kHonorEventIds[] = {
	"HONOR_EVENT_LOOT_INNOCENT", "HONOR_EVENT_AMBIENT_KILL", "HONOR_EVENT_AMBIENT_KO",
	"HONOR_EVENT_SCARE", "HONOR_EVENT_KILL_VERMIN", "HONOR_EVENT_KILL_FARM_ANIMAL",
	"HONOR_EVENT_KILL_HORSE", "HONOR_EVENT_STEAL_HORSE", "HONOR_EVENT_STEAL_DONKEY",
	"HONOR_EVENT_STEAL_MULE", "HONOR_EVENT_TRAMPLED_INNOCENT", "HONOR_EVENT_STEAL_WAGON",
	"HONOR_EVENT_ABANDON_ANIMALS", "HONOR_EVENT_ANIMAL_BLEEDOUT", "HONOR_EVENT_ANTAGONIZE",
	"HONOR_EVENT_THEFT", "HONOR_EVENT_INTERVENED", "HONOR_EVENT_WANTED_IN_CAMP",
	"HONOR_EVENT_DONATED_GAME", "HONOR_EVENT_ITEM_REQUEST", "HONOR_EVENT_LONG_ABSENCE"
};
static const int kHonorVanillaTiers[] = {
	-640, -480, -320, -160, -40, -20, -10, -5, -2, -1,
	0, 1, 2, 5, 10, 20, 40, 160, 640
};
static HonorActionTier g_honorActionTiers[19];
static int g_honorActionDisabledBits = 0;
static bool g_honorActionsLoaded = false;
static DWORD g_honorActionsNextReloadAt = 0;
static int g_honorActionsLastValue = 0;
static bool g_honorActionsHaveValue = false;

static bool honorActionBool(const std::string& text, bool* value) {
	if (text == "1" || text == "true") { *value = true; return true; }
	if (text == "0" || text == "false") { *value = false; return true; }
	return false;
}

static void honorActionsReload(DWORD now) {
	if (now < g_honorActionsNextReloadAt) return;
	g_honorActionsNextReloadAt = now + 2000;
	std::ifstream input(g_moduleDir + "\\honor_actions.csv");
	if (!input) { g_honorActionsLoaded = false; return; }
	std::string line;
	if (!std::getline(input, line) || line != "kind,id,enabled,amount") return;
	int disabled = 0, eventCount = 0, tierCount = 0;
	HonorActionTier replacement[19];
	while (std::getline(input, line)) {
		std::vector<std::string> fields; std::istringstream row(line); std::string field;
		while (std::getline(row, field, ',')) fields.push_back(field);
		if (fields.size() < 3 || fields.size() > 4) return;
		bool enabled = true; if (!honorActionBool(fields[2], &enabled)) return;
		if (fields[0] == "event") {
			if (eventCount >= 21 || fields[1] != kHonorEventIds[eventCount]) return;
			if (!enabled) disabled |= kHonorEventBits[eventCount];
			++eventCount;
		} else if (fields[0] == "tier") {
			if (tierCount >= 19 || fields.size() != 4) return;
			const std::string expected = "tier_" + std::string(kHonorVanillaTiers[tierCount] >= 0 ? "+" : "") + std::to_string(kHonorVanillaTiers[tierCount]);
			if (fields[1] != expected) return;
			char* end = nullptr; const long amount = std::strtol(fields[3].c_str(), &end, 10);
			if (!end || *end || amount < -10000 || amount > 10000) return;
			replacement[tierCount] = { kHonorVanillaTiers[tierCount], (int)amount, enabled };
			++tierCount;
		} else return;
	}
	if (eventCount != 21 || tierCount != 19) return;
	for (int i = 0; i < 19; ++i) g_honorActionTiers[i] = replacement[i];
	g_honorActionDisabledBits = disabled; g_honorActionsLoaded = true;
}

static void updateHonorActions(DWORD now) {
	honorActionsReload(now);
	if (!g_honorActionsLoaded) { g_honorActionsHaveValue = false; return; }
	constexpr int managedMask = 0x7FFFE7; // exact 21 audited bits; gaps remain Rockstar-owned
	UINT64* flags = getGlobalPtr(36616);
	*flags = (*flags & ~(UINT64)managedMask) | (UINT64)g_honorActionDisabledBits;
	int current = 0; if (!STAT_GET(joaat("HONOR_CURRENT"), &current)) return;
	if (!g_honorActionsHaveValue) { g_honorActionsLastValue = current; g_honorActionsHaveValue = true; return; }
	const int delta = current - g_honorActionsLastValue;
	for (int i = 0; i < 19; ++i) if (delta == g_honorActionTiers[i].vanilla) {
		const int replacement = g_honorActionTiers[i].enabled ? g_honorActionTiers[i].amount : 0;
		current = (std::max)(-320, (std::min)(320, g_honorActionsLastValue + replacement));
		if (current != g_honorActionsLastValue + delta) STAT_SET(joaat("HONOR_CURRENT"), current);
		break;
	}
	g_honorActionsLastValue = current;
}
